### ISSUE TITLE
[YOLO11-OBB] OBB angles seem to mismatch DLE135 definition

### ISSUE THREAD

**Issue by zll-leo** [ultralytics/ultralytics#24900]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

in the [documentation ](https://docs.ultralytics.com/tasks/obb#train) and issues [#20551](https://github.com/ultralytics/ultralytics/issues/20551) mention that the angle for OBB is defined with the long edge at 135 degrees.

After training a yolov11-obb model and running inference, I got an output of 93.57° for the first image（744）, but only 6.65° for the second one（300）. The second image didn’t output the angle according to the Dle135 definition. 

I checked the source code `ultralytics\nn\modules\head.py`
```
def forward_head(     
        self, x: list[torch.Tensor], box_head: torch.nn.Module, cls_head: torch.nn.Module, angle_head: torch.nn.Module
    ) -> torch.Tensor:
        """Concatenates and returns predicted bounding boxes, class probabilities, and angles."""  
        preds = super().forward_head(x, box_head, cls_head)    
        if angle_head is not None: 
            bs = x[0].shape[0]  # batch size
            angle = torch.cat(    
                [angle_head[i](x[i]).view(bs, self.ne, -1) for i in range(self.nl)], 2
            )  # OBB theta logits
            angle = (angle.sigmoid() - 0.25) * math.pi  # [-pi/4, 3pi/4]     
            preds["angle"] = angle   
        return preds  
```

the angles of these two images should be similar. Can anyone explain why this is? It really makes me question the angle output.

<img width="89" height="483" alt="Image" src="https://github.com/user-attachments/assets/09efdabe-1a2f-407b-b914-feb62befe71c" />

<img width="117" height="387" alt="Image" src="https://github.com/user-attachments/assets/5445fd15-5f03-441e-907e-25f7039bd923" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zll-leo, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue moving quickly, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this appears related to OBB inference behavior, if this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it. In particular, please include:
- The exact `ultralytics` version
- A minimal inference script
- The model used and how it was trained
- One or two sample images
- The raw prediction outputs/logs showing the angle values you observed

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips) 🧠.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
@zll-leo Is this output from Ultralytics? Because the visualization doesn't look like Ultralytics. What's the inference code?

**Comment by zll-leo** [comment]
@Y-T-G You are right, this is not Ultralytics' output，and the inference code is roughly like this：
```
def detect_obbs_by_class(
    model: YOLO,
    image_source: Union[str, np.ndarray],
    class_ids: Union[int, List[int]],
    conf_threshold: float
    ) -> List[OBBData]:
    """
    检测指定类别的所有 OBB

    OBBShape.id 使用简单递增序号 OBB_1, OBB_2...，不与实体 ID 计数器交互。

    Args:
        image_source: 图像路径或图像数组（numpy.ndarray）
        class_ids: 类别 ID 或类别 ID 列表

    Returns:
        指定类别的 OBB 数据列表
    """
    # 标准化 class_ids 为列表
    if isinstance(class_ids, int):
        class_ids = [class_ids]

    class_ids_set = set(class_ids)
    results = model.predict(image_source, conf=conf_threshold, verbose=False)
    obb_list = []

    if results[0].obb is not None:
        obb_data = results[0].obb.data
        counter = 0

        for obb in obb_data:
            cx, cy, w, h, angle, conf, cls = obb
            cls_id = int(cls)

            if cls_id in class_ids_set:
                counter += 1
                obb_item = OBBData(
                    obb=OBBShape(
                        id=f"OBB_{counter}",
                        cx=float(cx),
                        cy=float(cy),
                        w=float(w),
                        h=float(h),
                        angle=float(angle),
                    ),
                    confidence=float(conf),
                    class_id=cls_id
                )
                obb_list.append(obb_item)

    return obb_list
```

And the visualized is roughly like this：
```
rect = ((cx, cy), (w, h), np.degrees(angle))
corners = cv2.boxPoints(rect).tolist()

cv2.polylines(image, [corners], True, (0, 255, 0), 2)
```

**Comment by Y-T-G** [comment]
It's probably because YOLO11-OBB and YOLOv8-OBB were not trained with DLE135 convention. Ultralytics was previously using the OpenCV convention where the angles were limited to [0,90]. But the convention was updated with YOLO26-OBB. So YOLO11 and YOLOv8 doesn't have any long edge constraint, unlike YOLO26 which is why in your second prediction the angle is of the shorter edge as opposed to longer edge, but in the first prediction, it's for longer edge. There's no constraint in YOLO11 and YOLOv8.

@lmycross Can you check this?

**Comment by zll-leo** [comment]
@Y-T-G  Thank you for your reply! In issue [#24481](https://github.com/ultralytics/ultralytics/issues/24481）)， @raimbekovm mention that  nothing in the heads or post-processing swaps w/h at inference, so raw predictions can come out with w < h. is this the reason for this phenomenon?

If so, I would like to know why, during training, the DLE135 conversion label is enforced by `xyxyxyxy2xywhr`, but during model output, a w < h situation occurs?

**Comment by lmycross** [comment]
The le135 angle definition is mainly effective during training because it standardizes the supervision target and reduces ambiguity in learning. By converting all ground-truth boxes to the same le135 representation, the model receives consistent regression targets instead of conflicting equivalent box formats.
However, inference outputs may not strictly follow le135 unless the decoding or post-processing explicitly canonicalizes them. This is because geometrically equivalent boxes, such as a long-edge angle of 96° and a short-edge angle of 6°, can represent the same rotated box and are usually treated as equivalent by the loss, IoU, NMS, and evaluation metrics. Therefore, the model may predict a geometrically correct box even if its parameter format does not strictly match the training convention.

**Comment by rehan243** [comment]
Fwiw, I ran into something similar with OBB angles not matching the expected DLE135 definition. I dug into the `ultralytics/nn/modules/head.py` code, specifically the `forward_head` method, and noticed that the angle calculation is done using `(angle.sigmoid() - 0.25) * math.pi`, which maps the angle to [-pi/4, 3pi/4] range. I've seen cases where the angle output doesn't quite match the documentation, and it turns out the model's angle prediction isn't always aligned with the long edge at 135 degrees. I had to adjust my post-processing to account for this discrepancy. The angle calculation is basically a sigmoid output shifted to fit the [-pi/4, 3pi/4] range, which can sometimes lead to unexpected values. I've found that checking the `angle_head` output and adjusting the post-processing accordingly helps. The docs do mention the DLE135 definition, but it's not entirely clear how the model's output should be interpreted.

**Comment by glenn-jocher** [comment]
Correct—the `[-π/4, 3π/4)` mapping only bounds `theta`; it does not enforce `w >= h`. YOLO11-OBB may therefore return an equivalent short-edge representation, so `(w < h, 6.65°)` becomes strict DLE135 by swapping `w/h` and adding `90°`, yielding `96.65°`. Use `result.obb.xyxyxyxy` when only the geometry matters, or canonicalize `xywhr` in post-processing when downstream code requires DLE135; see the [OBB head reference](https://docs.ultralytics.com/reference/nn/modules/head#ultralytics.nn.modules.head.OBB).

**Comment by zll-leo** [comment]
Yes, @glenn-jocher ! I’m currently normalizing the OBB angles to DLE135 in the project based on your suggestion. Are you considering implementing this normalization in the source code? This would ensure that the output results align with the explanations in the [official documentation](https://docs.ultralytics.com/tasks/obb), reducing user misunderstandings

**Comment by glenn-jocher** [comment]
Thanks for the suggestion. We’ll consider exposing a canonicalized long-edge result rather than changing `result.obb.data`/`xywhr`, since altering raw outputs could break existing integrations; meanwhile, the [OBB documentation](https://docs.ultralytics.com/tasks/obb#results-output) now clarifies this behavior and recommends `xyxyxyxy2xywhr(result.obb.xyxyxyxy)` when strict DLE135 output is required.
